tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -325,6 +359,89 @@ class HLSLCallDiagnoseVisitor
     return true;
   }
 
+  bool VisitVarDecl(VarDecl *VD) {
+    QualType VarType = VD->getType();
+    if (const TemplateSpecializationType *TST =
+            dyn_cast<TemplateSpecializationType>(VarType.getTypePtr())) {
+      const TemplateDecl *TD = TST->getTemplateName().getAsTemplateDecl();
+      if (!TD)
+        return true;
+
+      // verify this is a rayquery decl
+      if (TD->getTemplatedDecl()->hasAttr<HLSLRayQueryObjectAttr>()) {
+        if (TST->getNumArgs() == 1) {
+          return true;
+        }
+        // now guaranteed 2 args
+        const TemplateArgument &Arg2 = TST->getArg(1);
+        Expr *Expr2 = Arg2.getAsExpr();
+        llvm::APSInt Arg2val;
+        Expr2->isIntegerConstantExpr(Arg2val, sema->getASTContext());
+
+        const ShaderModel *SM = hlsl::ShaderModel::GetByName(
+            sema->getLangOpts().HLSLProfile.c_str());
+
+        if (Arg2val.getZExtValue() != 0 && !SM->IsSMAtLeast(6, 9)) {
+          // if it's an integer literal, emit
+          // warn_hlsl_rayquery_flags_disallowed
+          if (Arg2.getKind() == TemplateArgument::Expression) {
+            if (auto *castExpr = dyn_cast<ImplicitCastExpr>(
+                    Arg2.getAsExpr()->IgnoreParens())) {
+              // Now check if the sub-expression is a DeclRefExpr
+              Expr *subExpr = castExpr->getSubExpr();
+              if (auto *IL = dyn_cast<IntegerLiteral>(subExpr))
+                sema->Diag(VD->getLocStart(),
+                           diag::warn_hlsl_rayquery_flags_disallowed);
+              return true;
+            }
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  bool VisitDeclRefExpr(DeclRefExpr *DRE) {
+    // Diagnose availability for referenced decl.
+    // Skip redundant availability diagnostics for the same Decl.
+    if (DiagnosedDecls.insert(DRE).second)
```

Comment:
The code between `VisitDeclRefExpr`, `GetAttrAndStoreDecl`, and `DiagnoseAvailability` looks a bit confused.
I think the intent was to put this `if (DiagnosedDecls.insert(DRE).second)` check inside `GetAttrAndStoreDecl`, then you don't need to condition the call `DiagnoseAvailability` here.